src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,475 @@
+
+        fieldsInStatelessChecksShouldBeImmutable.check(CHECKSTYLE_CHECKS);
+    }
+
+    /**
+     * Test to ensure classes with immutable fields are annotated with {@link StatelessCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testClassesWithImmutableFieldsShouldBeStateless() {
+        final ArchCondition<JavaClass> beSuppressedClass = new SuppressionArchCondition<>(
+            SUPPRESSED_CLASSES_FOR_STATELESS_CHECK_RULE, "be suppressed");
+
+        final ArchRule classesWithImmutableFieldsShouldBeStateless = classes()
+            .that(have(IMMUTABLE_FIELDS))
+            .and()
+            .haveSimpleNameEndingWith("Check")
+            .and()
+            .doNotHaveModifier(JavaModifier.ABSTRACT)
+            .should()
+            .beAnnotatedWith(StatelessCheck.class)
+            .orShould(beSuppressedClass);
+
+        classesWithImmutableFieldsShouldBeStateless.check(CHECKSTYLE_CHECKS);
+    }
+
+    /**
+     * Test to ensure classes with mutable fields are annotated with {@link FileStatefulCheck} or
+     * {@link GlobalStatefulCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testClassesWithMutableFieldsShouldBeStateful() {
+        final ArchCondition<JavaClass> beSuppressedClass = new SuppressionArchCondition<>(
+            SUPPRESSED_CLASSES_FOR_STATEFUL_CHECK_RULE, "be suppressed");
+
+        final ArchRule classesWithMutableFieldsShouldBeStateful = classes()
+            .that(doNot(have(IMMUTABLE_FIELDS)))
+            .and()
+            .haveSimpleNameEndingWith("Check")
+            .and()
+            .doNotHaveModifier(JavaModifier.ABSTRACT)
+            .should()
+            .beAnnotatedWith(FileStatefulCheck.class)
+            .orShould()
+            .beAnnotatedWith(GlobalStatefulCheck.class)
+            .orShould(beSuppressedClass);
+
+        classesWithMutableFieldsShouldBeStateful.check(CHECKSTYLE_CHECKS);
+    }
+
+    /**
+     * ArchCondition checking fields are immutable.
+     */
+    private static final class ImmutableFieldArchCondition extends ArchCondition<JavaField> {
+        private ImmutableFieldArchCondition() {
+            super("be among immutable types");
+        }
+
+        /**
+         * Whether a field is immutable or not.
+         *
+         * @param javaField field to check
+         * @return {@code true} if field is immutable
+         */
+        private static boolean isImmutable(JavaField javaField) {
+            final JavaClass rawType = javaField.getRawType();
+            boolean isImmutable = true;
+            if (!rawType.isEnum() && !ZERO_SIZE_ARRAY_FIELDS.contains(javaField.getFullName())) {
```

Comment:
Please extract this to a boolean, maybe something like:
`final boolean isEnumConstantOrEmptyArray = rawType.isEnum() || ZERO_SIZE_ARRAY_FIELDS.contains(javaField.getFullName()`